### PR TITLE
[aggregator] keep metric type during the aggregation

### PR DESCRIPTION
**Pull request by gediminasgu**
**What this PR does / why we need it**:
Ensures that metric type will be kept after the aggregation process in the aggregator.

**Special notes for your reviewer**:
Metric type is sent in the same way as graphite type - using a special tag. Later that tag is removed before sending to dbnode.

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:
```documentation-note
NONE
```

### CODE REVIEW

File: scripts/docker-integration-tests/aggregator/m3coordinator.yml [hunk 1]
```diff
@@ -77,3 +77,5 @@ ingest:
       retry:
         maxBackoff: 10s
         jitter: true
+
+storeMetricsType: true
```
Comment (linasm) [inline scripts/docker-integration-tests/aggregator/m3coordinator.yml:81]: Nit: newline.

File: src/cmd/services/m3coordinator/ingest/write.go [hunk 2]
```diff
@@ -471,6 +471,13 @@ func (d *downsamplerAndWriter) writeAggregatedBatch(
 			appender.AddTag(downsample.MetricsOptionIDSchemeTagName,
 				downsample.GraphiteIDSchemeTagValue)
 		}
+		// If prom type is set then send it to aggregator too to do not loose it.
+		// Currently it's passed as a tag and later filtered out when series will back to
+		// the coordinator.
+		// Also, see note above.
+		if value.Attributes.PromType != ts.PromMetricTypeUnknown {
+			appender.AddTag(downsample.PromTypeTagValue, []byte{byte(value.Attributes.PromType)})
```
Comment (linasm) [inline src/cmd/services/m3coordinator/ingest/write.go:479]: I don't feel very confident with storing non-printable characters (such as `0` byte) in tag value bytes.

File: src/query/api/v1/handler/prometheus/remote/write.go [hunk 3]
```diff
@@ -493,6 +493,31 @@ func (h *PromWriteHandler) parseRequest(
 		}
 	}
 
+	if promType := r.Header.Get(headers.PromTypeHeader); promType != "" {
```
Comment (linasm) [inline src/query/api/v1/handler/prometheus/remote/write.go:496]: What is the purpose of adding this header?

File: src/query/api/v1/handler/prometheus/remote/write.go [hunk 4]
```diff
@@ -493,6 +493,31 @@ func (h *PromWriteHandler) parseRequest(
...
+		case "gauge-histogram":
+			tp = prompb.MetricType_GAUGE_HISTOGRAM
+		case "histogram":
+			tp = prompb.MetricType_HISTOGRAM
+		case "info":
+			tp = prompb.MetricType_INFO
+		case "stateset":
+			tp = prompb.MetricType_STATESET
+		case "summary":
+			tp = prompb.MetricType_SUMMARY
```
Comment (linasm) [inline src/query/api/v1/handler/prometheus/remote/write.go:512]: Nit: might want to replace this `switch` with a `map`.

File: src/query/storage/m3/storage.go [hunk 5]
```diff
@@ -719,6 +720,25 @@ func (s *m3storage) writeSingle(
 
 	namespaceID := namespace.NamespaceID()
 	session := namespace.Session()
+
+	annot := query.Annotation()
+	if query.Options().Type != ts.PromMetricTypeUnknown && (annot == nil || len(annot) == 0) {
```
Comment (linasm) [inline src/query/storage/m3/storage.go:725]: `len(annot) == 0` should cover the `nil` case.

File: src/query/storage/types.go [hunk 6]
```diff
@@ -245,6 +245,7 @@ type WriteQueryOptions struct {
 	Datapoints ts.Datapoints
 	Unit       xtime.Unit
 	Annotation []byte
+	Type       ts.PromMetricType
```
Comment (linasm) [inline src/query/storage/types.go:248]: Why is this needed, isn't the type already present in the annotation when this struct is created? If there are some cases when it isn't, I think we should fix them instead. Having two competing fields for the same purpose will only cause trouble.

Comment (linasm) [reply src/query/storage/types.go:248]: Also, where does this field get assigned?

File: src/query/api/v1/handler/prometheus/remote/write.go (hunk 4 above)
Comment (linasm) [reply src/query/api/v1/handler/prometheus/remote/write.go:512]: Also I am unable to find any test coverage for this logics.

**Review by linasm — CHANGES_REQUESTED** [review]
What was the reason to use a synthetic tag to pass the metrics type around (as opposed to adding a new explicit field for it)?
Not only this feels hacky, but also it would probably require organizing a two phase deployment to prevent polluting the tags of clients metrics (as it would not be possible to upgrade all the write coordinators simultaneously).

File: src/query/storage/types.go (hunk 6 above)
Comment (gediminasgu) [reply src/query/storage/types.go:248]: We assign this field when we receive data from the aggregator. Aggregator doesn't store any information about annotations: https://github.com/m3db/m3/blob/gg/aggregator-keep-metric-type/src/cmd/services/m3coordinator/ingest/m3msg/ingest.go#L222

File: src/aggregator/aggregator/generic_elem.go [hunk 7]
```diff
@@ -537,9 +537,9 @@ func (e *GenericElem) processValueWithAggregationLock(
 			for _, point := range toFlush {
 				switch e.idPrefixSuffixType {
 				case NoPrefixNoSuffix:
-					flushLocalFn(nil, e.id, nil, point.TimeNanos, point.Value, e.sp)
+					flushLocalFn(nil, e.id, metric.GaugeType, nil, point.TimeNanos, point.Value, e.sp)
 				case WithPrefixWithSuffix:
-					flushLocalFn(e.FullPrefix(e.opts), e.id, e.TypeStringFor(e.aggTypesOpts, aggType),
+					flushLocalFn(e.FullPrefix(e.opts), e.id, metric.GaugeType, e.TypeStringFor(e.aggTypesOpts, aggType),
```
Comment (linasm) [inline src/aggregator/aggregator/generic_elem.go:542]: Why is it a `Gauge`? Is the type not available here? If so, some comment might be useful.

File: src/query/api/v1/handler/prometheus/remote/write_test.go [hunk 8]
```diff
@@ -86,6 +86,31 @@ func TestPromWriteParsing(t *testing.T) {
 	require.Equal(t, ingest.WriteOptions{}, r.Options)
 }
 
+func TestMetricTypeHeader(t *testing.T) {
```
Comment (linasm) [inline src/query/api/v1/handler/prometheus/remote/write_test.go:89]: Can we use a test-table to cover all types?

File: src/x/headers/headers.go [hunk 9]
```diff
@@ -43,6 +43,10 @@ const (
 	// Valid values are "unaggregated" or "aggregated".
 	MetricsTypeHeader = M3HeaderPrefix + "Metrics-Type"
 
+	// PromTypeHeader sets the prometheus metric type. Valid values are
+	// "counter", "gauge", etc.
```
Comment (linasm) [inline src/x/headers/headers.go:47]: Nit: add a link to the code that has those values listed so that there is no need for "etc".

File: src/aggregator/aggregator/generic_elem.go (hunk 7 above)
Comment (gediminasgu) [reply src/aggregator/aggregator/generic_elem.go:542]: This is a base code for the code generation. By default, we set metric-type to Gauge (as most of them are) and just for Counters we replace it to Counter: https://github.com/m3db/m3/blob/gg/aggregator-keep-metric-type/src/aggregator/generated-source-files.mk#L16

File: src/cmd/services/m3coordinator/ingest/m3msg/ingest.go [hunk 10]
```diff
@@ -139,19 +144,21 @@ func (i *Ingester) Ingest(
...
+	tagOpts          models.TagOptions
+	p                pool.ObjectPool
+	m                ingestMetrics
+	logger           *zap.Logger
+	sampler          *sampler.Sampler
+	attemptFn        retry.Fn
+	ingestFn         func()
+	storeMetricsType bool
 
 	c           context.Context
```
Comment (linasm) [inline src/cmd/services/m3coordinator/ingest/m3msg/ingest.go:159]: Nit: I know this is prior code, but maybe you could take an opportunity and change all those single character field names here to something comprehensible?

File: src/cmd/services/m3coordinator/ingest/m3msg/ingest.go [hunk 11]
```diff
@@ -223,7 +231,39 @@ func (op *ingestOp) resetWriteQuery() error {
...
+	if tp == ts.PromMetricTypeUnknown {
+		return nil, nil
+	}
+
+	atp, err := storage.PromMetricTypeToAnnotationPayloadType(tp)
+	if err != nil {
+		return nil, err
+	}
+
+	annotationPayload := annotation.Payload{MetricType: atp}
```
Comment (linasm) [inline src/cmd/services/m3coordinator/ingest/m3msg/ingest.go:257]: This is missing the `HandleValueResets` field which should be set to `true` for counters (which indicates that some unit test should be added for such case).
I suggest to fix it by reverting the original `SeriesAttributesToAnnotationPayload` function and modifying to take `PromType` and `HandleValueResets` as arguments. Then you should be able to reuse it here directly, and avoid duplication of `annotation.Payload` construction code with all the consequences.

**Review by linasm — CHANGES_REQUESTED** [review]

File: src/cmd/services/m3coordinator/ingest/m3msg/ingest.go [hunk 12]
```diff
@@ -223,15 +230,53 @@ func (op *ingestOp) resetWriteQuery() error {
...
+	}
+
+	handleValueResets := false
+	if tp == ts.PromMetricTypeCounter {
+		handleValueResets = true
+	}
+
+	annotationPayload, err := storage.SeriesAttributesToAnnotationPayload(ts.SeriesAttributes{
+		PromType:          tp,
+		HandleValueResets: handleValueResets,
```
Comment (linasm) [inline src/cmd/services/m3coordinator/ingest/m3msg/ingest.go:258]: It would be cleaner if `SeriesAttributesToAnnotationPayload` now took those two individual arguments and not the `ts.SeriesAttributes` (which probably does not fit well in this part of code). Then if some third field is added, the compiler will check that we need to pass it here as well. With struct, we lose that safety.

**Review by linasm — COMMENTED** [review]
Had a couple of last nits.
From there, I think it would be great if someone with more context on `m3aggregator` could have a look at this (@shreyassrivatsan and/or @ryanhall07, maybe?)

**Review by shreyassrivatsan — APPROVED** [review]
Looks good. Added a couple of questions..

File: src/cmd/services/m3coordinator/ingest/write.go [hunk 13]
```diff
@@ -495,14 +495,24 @@ func (d *downsamplerAndWriter) writeAggregatedBatch(
...
 		for _, dp := range value.Datapoints {
-			switch value.Attributes.M3Type {
-			case ts.M3MetricTypeGauge:
-				err = result.SamplesAppender.AppendGaugeTimedSample(dp.Timestamp, dp.Value)
-			case ts.M3MetricTypeCounter:
-				err = result.SamplesAppender.AppendCounterTimedSample(dp.Timestamp, int64(dp.Value))
-			case ts.M3MetricTypeTimer:
-				err = result.SamplesAppender.AppendTimerTimedSample(dp.Timestamp, dp.Value)
+			if value.Attributes.PromType != ts.PromMetricTypeUnknown {
+				switch value.Attributes.PromType {
```
[the remark this answers is not in the document]
Comment (gediminasgu) [reply src/cmd/services/m3coordinator/ingest/write.go:499]: That's correct: https://prometheus.io/docs/concepts/metric_types/

File: src/cmd/services/m3coordinator/ingest/write.go [hunk 14]
```diff
@@ -495,14 +495,24 @@ func (d *downsamplerAndWriter) writeAggregatedBatch(
...
+				case ts.PromMetricTypeCounter:
+					err = result.SamplesAppender.AppendCounterTimedSample(dp.Timestamp, int64(dp.Value))
+				default:
```
[the remark this answers is not in the document]
Comment (gediminasgu) [reply src/cmd/services/m3coordinator/ingest/write.go:502]: Actually, for now, we care only about counters. If there will be additional logic for other metric types, then the pipeline will need a review accordingly.

**Review by linasm — APPROVED** [review]
